src/main/java/object/moverule/CannonRule.java

```diff
@@ -0,0 +1,83 @@
+package object.moverule;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import object.coordinate.Position;
+import object.coordinate.Path;
+import object.coordinate.RelativePosition;
+import object.piece.Piece;
+import object.piece.PieceType;
+import object.piece.Team;
+
+public class CannonRule extends MoveRule {
+
+    @Override
+    public Path getLegalRoute(Position startPosition, Position endPosition, Team team) {
+        Position maxPosition = Position.getAbsoluteBigPositionBetween(startPosition, endPosition);
+
+        if (startPosition.isSameColumn(endPosition)) {
+            if (maxPosition.equals(startPosition)) {
+                return generateStraightRoute(startPosition, endPosition, new RelativePosition(-1, 0));
+            }
+            return generateStraightRoute(startPosition, endPosition, new RelativePosition(1, 0));
+        }
+
+        if (startPosition.isSameRow(endPosition)) {
+            if (maxPosition.equals(startPosition)) {
+                return generateStraightRoute(startPosition, endPosition, new RelativePosition(0, -1));
+            }
+            return generateStraightRoute(startPosition, endPosition, new RelativePosition(0, 1));
+        }
+
+        throw new IllegalArgumentException(MoveRule.INVALID_POSITION);
+    }
+
+    @Override
+    public boolean isAbleToThrough(Path legalPath, List<Piece> piecesOnBoard, Team team) {
+        // 규칙: 포는 목적지를 포함하지 않은 이동 경로에 무조건 포가 아닌 기물과 1회 충돌해야 함.
+        Position destination = legalPath.getLast();
+
+        List<Piece> piecesOnRoute = piecesOnBoard.stream()
+                .filter(piece -> legalPath.contains(piece.getPosition()))
+                .filter(piece -> !piece.isSamePosition(destination))
+                .toList();
+
+        if (piecesOnRoute.size() != 1) {
+            return false;
+        }
+
+        if (piecesOnRoute.getFirst().isSameType(getPieceType())) {
+            return false;
+        }
+
+        // 목적지에 기물이 있다면
+        Optional<Piece> pieceOnDestination = piecesOnBoard.stream()
+                .filter(piece -> piece.isSamePosition(destination))
+                .findFirst();
+
+        if (pieceOnDestination.isPresent() && pieceOnDestination.get().isSameTeam(team)) {
+            return false;
+        }
+
+        if (pieceOnDestination.isPresent() && pieceOnDestination.get().isSameType(getPieceType())) {
+            return false;
+        }
```

Comment:
여기에 대한 테스트가 없네요. 그리고 아래와 같이 조건문들을 합칠 수 있지 않을까요?

```suggestion
        if (pieceOnDestination.isPresent() && (pieceOnDestination.get().isSameTeam(team) || pieceOnDestination.get().isSameType(getPieceType()))) {
            return false;
        }
```

if문을 사용할 필요없이 바로 조건의 결과를 리턴해도 될 것 같군요

Reply:
조건문을 합치는 것도 좋은 방법이지만, 코드 라인을 단순히 줄이는 것보다는 절차적인 사고를 코드에 녹여내서 최대한 흐름을 이해하기 쉽게 만드는 것도 중요하다고 생각합니다. 베루스가 제안해주신 방법으로도 코드를 이해할 수 있지만, 제 생각에는 조건문의 길이가 너무 길고 논리 연산자와 괄호가 합쳐져 있어 크게 가독성이 좋아지는 느낌은 받지 못하는 것 같습니다. 

두 개의 논리 흐름을 합쳐서 작성했을 때의 장점과, 분리해서 작성하는 것의 단점을 설명해주실 수 있으신가요?

Reply:
두개의 논리 흐름을 합치면 관심사를 한번에 볼 수 있다고 생각합니다. false인 케이스를 하나 보고 이후에 또 본다면 false가 끝나지 않았기에 다음 조건도 false 또는 true가 올 수 있구나 생각할 것 같아요. 하지만 false 케이스를 한번에 끝낸다면 다음 조건문을 만날 때는 false 케이스에 대한 생각을 지우고 코드를 볼 수 있다고 생각합니다.

조건문을 합쳐도 절차적인 사고를 충분히 녹여낼 수 있다고 생각합니다. 가독성이 부족하다면 좋은 이름의 메서드 또는 변수 추출을 해보는건 어떨까요?

Reply:
그렇네요. 설득되었습니다. false인 케이스들을 하나로 합치면 if문 내의 로직은 복잡해질 수 있어도, false를 반환한다 라는 관심사 측면에선 가독성이 높아질 것 같네요.

메서드를 좋은 이름으로 추출하는 것도 가독성을 올릴 좋은 방법일 것 같습니다 ㅎㅎ
피드백 감사합니다 :)